### PR TITLE
Rga common analysis functions :: FiducialFilter

### PR DESCRIPTION
**Pull request by tbhayward**
Created new filter algorithm RGAFiducialFilter with support for RGA "pass-2" DC, CAL, CVT and FT fiducial cuts for relevant tracks. Filter checks each track in REC::Particle and applies the appropriate detector cuts. Default "RGA approved" values are stored in the associated Config.yaml for the DC region edge cuts, calorimeter lv & lw cuts, holes in forward tagger and gaps between sectors in CVT. These values can be changed by modifying the Config.yaml.

Strictness (default = 1) can be set as 1,2,3 by the user to increase tightness of calorimeter cuts. Future updates for run-by-run cuts on dead PMTs, regions of DC and CVT, etc. are planned.

Validator creates plots of FT electrons and photons before and after cuts, DC inbending and outbending tracks for region 1, 2 and 3 before and after cuts, CAL lv and lw distributions before and after cuts and CVT theta vs phi for all hadrons before and after cuts.

### CODE REVIEW

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+//    !!! TO DO: implement run-by-run data/MC matching for missing PMTs, etc.
+//
+// * Central Detector cut on areas between 3 sectors and require tracks inside detector (edge > 0);
+//    for all charged hadrons
+// 
+// * Drift chamber cut on region 1 > 3 cm, region 2 > 3, region3 > 10; inbending tracks also require
+//    region 1 > 10, region 2 > 10 for theta < 10; cuts for all charged leptons and hadrons
+
+namespace iguana::clas12 {
+static bool banklist_has(hipo::banklist& banks, const char* name) {
```
Comment (c-dilks) [inline src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc:33]: The CI tests aren't running for this; adding `REGISTER_IGUANA_ALGORITHM` here should fix that:

```suggestion
namespace iguana::clas12 {

REGISTER_IGUANA_ALGORITHM(RGAFiducialFilter, "clas12::RGAFiducialFilter");

static bool banklist_has(hipo::banklist& banks, const char* name) {
```

I'll take a closer look at this PR later this week; for now, let's see if the CI tests pass (they run on every new commit you push).

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Config.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+clas12::RGAFiducialFilter:
+  calorimeter:
+    # Default strictness if not overridden programmatically via SetStrictness(1|2|3)
+    # s=1 -> lv,lw >=  9.0 cm
+    # s=2 -> lv,lw >= 13.5 cm
+    # s=3 -> lv,lw >= 18.0 cm
+    #
+    # s=1 loosely recommended for spin-asymmetries
+    strictness: [1] # default
```
[the remark this answers is not in the document]
Comment (tbhayward) [reply src/iguana/algorithms/clas12/RGAFiducialFilter/Config.yaml:49]: switched strictness to scalar

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Config.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      [ 1.60, -8.42,  9.89,
+        1.60, -9.89, -5.33,
+        2.30, -6.15, -13.00,
+        2.00,  3.70, -6.50 ]
+
+  cvt:
+    # Layers to apply the edge>edge_min test to (all). Missing layers are treated as pass.
+    edge_layers: [1, 3, 5, 7, 12]
+    # edge > 0 to ensure tracks inside CVT
+    edge_min: [0.0]
```
[the remark this answers is not in the document]
Comment (tbhayward) [reply src/iguana/algorithms/clas12/RGAFiducialFilter/Config.yaml]: changed to scalar

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+    double in_large_e1 = 3.0,  in_large_e2 = 3.0,  in_large_e3 = 10.0;
+    // outbending (any theta)
+    double out_e1      = 3.0,  out_e2      = 3.0,  out_e3      = 10.0;
+  };
+  DCParams g_dc;
+  } // namespace
+
+
+  // YAML loading (required)
+  void RGAFiducialFilter::LoadConfigFromYAML() {
```
[the remark this answers is not in the document]
Comment (c-dilks) [reply src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc]: It looks like you are still relying on your custom YAML parsing functions (named `RGAFID_*`). Your `Config.yaml` file is fully supported by our `iguana::YAMLReader`, which lies under the hood of functions like `Algorithm::GetOptionScalar`; please parse configuration with _these_ functions, rather than re-implementing or wrapping them.

```cpp
void RGAFiducialFilter::Start(hipo::banklist& banks)
{
  // ...
  ParseYAMLConfig();
  m_cvt.edge_layers = GetOptionVector<int>("cvt.edge_layers", {"cvt", "edge_layers"});
  m_cvt.edge_min    = GetOptionScalar<double>("cvt.edge_min", {"cvt", "edge_min"});
  // ...
}
```
See other algorithms if you need more examples.
See also `src/iguana/tests/include/TestConfig.h`.

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+  // One edge value per required layer for this track 
+  std::map<int, double> edge_at_layer;
+
+  double x12 = 0.0, y12 = 0.0;
+  bool saw12 = false;
+
+  for (int i = 0; i < n; ++i) {
+    if (traj.getInt("pindex", i)   != pindex) continue;
+    if (traj.getInt("detector", i) != 5     ) continue; // CVT
```
Comment (c-dilks) [inline src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc:233]: don't hard-code detector numbers, we have [`enum DetectorType`](https://github.com/JeffersonLab/iguana/blob/7744c9d930cf2404931c94123801b86eee1e0c0a/src/iguana/algorithms/TypeDefs.h#L136-L165) to handle this; you'll probably need to `#include` the `TypeDefs.h` header file to use this.
```suggestion
    if (traj.getInt("detector", i) != DetectorType::CVT     ) continue; // CVT
```

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+  const double theta = std::atan2(rho, (pz==0.0 ? 1e-12 : pz)) * (180.0 / kPI);
+
+  double e1=0.0, e2=0.0, e3=0.0;
+  bool   saw_dc = false;
+
+  const auto& traj = *trajBank;
+  const int n = traj.getRows();
+  for (int i=0; i<n; ++i) {
+    if (traj.getInt("pindex", i) != pindex) continue;
+    if (traj.getInt("detector", i) != 6)    continue; // DC
```
Comment (c-dilks) [inline src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc:305]: use `DetectorType`
```suggestion
    if (traj.getInt("detector", i) != DetectorType::DC)    continue; // DC
```

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+    if (!b) return false;
+    const int n = b->getRows();
+    for (int i=0;i<n;++i) if (b->getInt("pindex", i) == track_index) return true;
+    return false;
+  };
+
+  const bool hasCal = has_assoc(calBank);
+  const bool hasFT  = has_assoc(ftBank);
+  const bool hasCVT = traj_has_detector(trajBank, track_index, 5);
+  const bool hasDC  = traj_has_detector(trajBank, track_index, 6);
```
Comment (c-dilks) [inline src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc:350]: use `DetectorType`
```suggestion
  const bool hasCVT = traj_has_detector(trajBank, track_index, DetectorType::CVT);
  const bool hasDC  = traj_has_detector(trajBank, track_index, DetectorType::DC);
```

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+    if (it == edge_at_layer.end()) continue;            // missing layer -> pass
+    if (!(it->second > m_cvt.edge_min)) {               // fail if edge <= edge_min
+      return false;
+    }
+  }
+
+  // Phi veto from YAML 
+  if (saw12 && !m_cvt.phi_forbidden_deg.empty()) {
+    constexpr double kPI = 3.141593;
+    double phi = std::atan2(y12, x12) * (180.0 / kPI);
```
Comment (c-dilks) [inline src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc:263]: use `M_PI`
```suggestion
    double phi = std::atan2(y12, x12) * (180.0 / M_PI);
```

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+  const bool electron_out = (torus == 1.0f);
+  const bool particle_inb = (electron_out ? isPos : isNeg);
+  const bool particle_out = !particle_inb;
+
+  const double px = particleBank.getFloat("px", pindex);
+  const double py = particleBank.getFloat("py", pindex);
+  const double pz = particleBank.getFloat("pz", pindex);
+  const double rho = std::hypot(px, py);
+  const double kPI = 3.141593;
+  const double theta = std::atan2(rho, (pz==0.0 ? 1e-12 : pz)) * (180.0 / kPI);
```
Comment (c-dilks) [inline src/iguana/algorithms/clas12/RGAFiducialFilter/Algorithm.cc:296]: use `M_PI`
```suggestion
  const double theta = std::atan2(rho, (pz==0.0 ? 1e-12 : pz)) * (180.0 / M_PI);
```

File: src/iguana/algorithms/clas12/RGAFiducialFilter/Config.yaml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        2.00,  3.70, -6.50 ]
+
+  cvt:
+    # Layers to apply the edge>edge_min test to (all). Missing layers are treated as pass.
+    edge_layers: [1, 3, 5, 7, 12]
+    # edge > 0 to ensure tracks inside CVT
+    edge_min: 0.0
+    # Forbidden phi wedges in degrees (open intervals).
+    # Example: (25,40), (143,158), (265,280)
+    phi_forbidden_deg: [25, 40, 143, 158, 265, 280]
```
[the remark this answers is not in the document]
Comment (c-dilks) [reply src/iguana/algorithms/clas12/RGAFiducialFilter/Config.yaml:29]: https://github.com/JeffersonLab/iguana/issues/370

**Comment by c-dilks** [comment]
I updated your branch with respect to `main`, to resolve a conflict coming from #368, which I recently merged. The merge commit on your branch is 46ea9927c2a1b36c730177870ef8baac45304462. You will need to run `git pull` to pick up this change in your local repository.
